config.org
My own personal emacs config
* my own navigation/kill/yank/insert/etc. functions
 #+BEGIN_SRC emacs-lisp
 ;; functions
 (defun backward-kill-line ()
   "Kill a backward line"
   (interactive)
   (kill-line -1)
   )

 (defun comment-or-uncomment-line-or-region ()
   "Comments or uncomments the current line or region."
   (interactive)
   (if (region-active-p)
       (comment-or-uncomment-region (region-beginning) (region-end))
     (comment-or-uncomment-region (line-beginning-position) (line-end-position))
     )
   )

 (defun kill-line-or-region ()
   "kills current line or region."
   (interactive)
   (if (region-active-p)
       (kill-region (region-beginning)(region-end))
     (kill-region (line-beginning-position) (line-end-position))
     )
   )

 (defun my-insert-date (prefix)
     "Insert the current date. With prefix-argument, use ISO format. With
    two prefix arguments, write out the day and month name."
     (interactive "P")
     (let ((format (cond
                    ((not prefix) "%Y-%m-%d %H:%M")
                    ((equal prefix '(4)) "%Y-%m-%d")
                    ((equal prefix '(16)) "%A, %d. %B %Y")))
           )
       (insert (format-time-string format))))

 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;               insert a file name
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

 (defun my-insert-file-name (filename &optional args)
     "Insert name of file FILENAME into buffer after point.

   Prefixed with \\[universal-argument], expand the file name to
   its fully canocalized path.  See `expand-file-name'.

   Prefixed with \\[negative-argument], use relative path to file
   name from current directory, `default-directory'.  See
   `file-relative-name'.

   The default with no prefix is to insert the file name exactly as
   it appears in the minibuffer prompt."
     ;; Based on insert-file in Emacs -- ashawley 20080926
     (interactive "*fInsert file name: \nP")
     (cond ((eq '- args)
            (insert (file-relative-name filename)))
           ((not (null args))
            (insert (expand-file-name filename)))
           (t
            (insert filename))))

 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;;;;;;;;;;Cleanup commands;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

 (defun untabify-buffer ()
   "Untabify current buffer"
   (interactive)
   (untabify (point-min) (point-max)))

 (defun indent-buffer()
   "indent whole buffer"
   (interactive)
   (delete-trailing-whitespace)
   (indent-region (point-min) (point-max) nil)
   (untabify (point-min) (point-max)))

 (defun cleanup-buffer ()
   "Perform a bunch of operations on the whitespace content of a buffer."
   (interactive)
   (indent-buffer)
   (untabify-buffer)
   (delete-trailing-whitespace))

#+END_SRC
* global keybindings
** Navigation
#+BEGIN_SRC emacs-lisp
  (define-key input-decode-map [?\C-m] [C-m])
  (global-set-key (kbd "C-t") 'previous-line)
  (global-set-key (kbd "C-w") 'next-line)
  (global-set-key (kbd "C-n") 'forward-char)
  (global-set-key (kbd "<C-m>") 'backward-char)
  (global-set-key (kbd "C-S-t") 'scroll-down)
  (global-set-key (kbd "C-S-w") 'scroll-up)
  (global-set-key (kbd "C-S-n") 'forward-word)
  (global-set-key (kbd "C-S-m") 'backward-word)
  (global-set-key (kbd "C-x C-n") 'end-of-line)
  (global-set-key (kbd "C-x <C-m>") 'beginning-of-line)
  (global-set-key (kbd "C-x C-w") 'end-of-buffer)
  (global-set-key (kbd "C-x C-t") 'beginning-of-buffer)

  (global-set-key (kbd "M-d") 'other-window)
  (global-set-key (kbd "C-o") 'split-window-right)
  (global-set-key (kbd "C-S-o") 'split-window-below)
  (global-set-key (kbd "C-x C-o") 'delete-window)
  (global-set-key (kbd "C-k") 'kill-buffer-and-window)

  (global-set-key (kbd "C-b") 'switch-to-next-buffer)
  (global-set-key (kbd "C-x C-b") 'switch-to-buffer)
  (global-set-key (kbd "M-b") 'switch-to-prev-buffer)
#+END_SRC
** help
#+BEGIN_SRC emacs-lisp

#+END_SRC
** Killing
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "M-t") 'backward-kill-line)
(global-set-key (kbd "M-w") 'kill-line)
(global-set-key (kbd "M-n") 'delete-forward-char)
(global-set-key (kbd "M-m") 'delete-backward-char)
(global-set-key (kbd "M-S-t") 'previous-line)
(global-set-key (kbd "M-S-w") 'next-line)
(global-set-key (kbd "M-N") 'kill-word)
(global-set-key (kbd "M-M") 'backward-kill-word)

#+END_SRC
** Copy, Cut, Paste, Undo
#+BEGIN_SRC emacs-lisp
  (global-set-key (kbd "C-s--") 'copy-region-as-kill)
  (global-set-key (kbd "C-=") 'yank)
  ;; (global-set-key (kbd "C-s-=") 'yank-pop)
  (global-set-key (kbd "C--") 'kill-line-or-region)
  (global-set-key (kbd "C-z") 'undo)
#+END_SRC
** Keyboard quit
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "M-q") 'keyboard-quit)
#+END_SRC
** Commenting
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "C-/") 'comment-or-uncomment-line-or-region)
#+END_SRC
** my custom functions
#+BEGIN_SRC emacs-lisp
(defun my-increment-number-decimal (&optional arg)
  "Increment the number forward from point by 'arg'."
  (interactive "p*")
  (save-excursion
    (save-match-data
      (let (inc-by field-width answer)
        (setq inc-by (if arg arg 2))
        (skip-chars-backward "0123456789")
        (when (re-search-forward "[0-9]+" nil t)
          (setq field-width (- (match-end 0) (match-beginning 0)))
          (setq answer (+ (string-to-number (match-string 0) 10) inc-by))
          (when (< answer 0)
            (setq answer (+ (expt 10 field-width) answer)))
          (replace-match (format (concat "%0" (int-to-string field-width) "d")
                                 answer)))))))

(defun my-decrement-number-decimal (&optional arg)
  (interactive "p*")
  (my-increment-number-decimal (if arg (- arg) -1)))

#+END_SRC
** unset keybindings
#+BEGIN_SRC emacs-lisp
  (global-unset-key (kbd "C-c C-s"))
#+END_SRC
** my kbd macros
#+BEGIN_SRC emacs-lisp

(fset 'start_hcube_venv_and_test
   [?\M-x ?f ?i ?n ?d ?  ?f ?i ?l ?e return ?\C-x ?\C-m ?\M-w ?/ ?h ?o ?m ?e ?/ ?n ?i ?c ?k ?/ ?p ?y ?t ?h ?o ?n ?_ ?s ?c ?r ?i ?p ?t ?s ?/ ?h ?o ?l ?o ?c ?u ?b ?e ?/ ?t ?e ?s ?t ?. ?p ?y return ?\M-x ?p ?y ?v ?e ?n ?v ?- ?w ?o ?r ?k ?o ?n return return])

(fset 'email-region
   (lambda (&optional arg) "Keyboard macro." (interactive "p") (kmacro-exec-ring-item (quote ([75497517 134217848 109 117 52 101 45 99 111 tab 109 112 111 115 101 tab tab 110 101 119 return 24 23 67108925 24 20 24 20 23 24 14] 0 "%d")) arg)))

(global-set-key (kbd "C-x C-k v") 'start_hcube_venv_and_test)
(global-set-key (kbd "C-x C-k m") 'email-region)
#+END_SRC

* shell commands to run at startup
#+BEGIN_SRC sh
setxkbmap -option ctrl:nocaps
#+END_SRC
* Emacs appearece menu,toolbar,scrollbar
#+BEGIN_SRC emacs-lisp
     (menu-bar-mode -1)
     (toggle-scroll-bar -1)
     (tool-bar-mode -1)
     (tooltip-mode -1)
     (scroll-bar-mode -1)
     (show-paren-mode t)
   (setq line-number-mode t)                     ;start in line-number-mode
   (setq column-number-mode t)                   ;start in column-number-mode
   (use-package rainbow-delimiters
   :ensure t
   :config
   (add-hook 'prog-mode-hook 'rainbow-delimiters-mode))


   (use-package highlight-indent-guides
     :ensure t
     :custom
     (highlight-indent-guides-method 'character)
     (highlight-indent-guides-responsive 'top)
     (highlight-indent-guides-auto-enabled nil)
     :custom-face
       (highlight-indent-guides-character-face ((t (:foreground ,"#2f4f4f"))))
       (highlight-indent-guides-top-character-face ((t (:foreground ,"#4682b4"))))
       (highlight-indentation-face ((t (:foreground ,"#2f4f4f"))))

     :hook
     (prog-mode . highlight-indent-guides-mode))
 #+END_SRC
* Emacs global settings
#+BEGIN_SRC emacs-lisp
  (tooltip-mode nil)                      ; one-line help text in the echo area
  (electric-indent-mode 1)

  ;;dont have to use "yes" or "no" instead use "y" or "n"
  (defalias 'yes-or-no-p 'y-or-n-p)

  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
  ;;;;;;;;;;;;;;;; emacs autocomplete with tab;;;;;;;;;;;
  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
  (add-hook
   'eshell-mode-hook
   (lambda ()
     (setq pcomplete-cycle-completions nil)))

  (autoload 'bash-completion-dynamic-complete
     "bash-completion"
     "BASH completion hook")
   (add-hook 'shell-dynamic-complete-functions
     'bash-completion-dynamic-complete)

  ;;;; sudo completion
  (defun pcomplete/sudo ()
    "Completion rules for the `sudo' command."
    (let ((pcomplete-ignore-case t))
      (pcomplete-here (funcall pcomplete-command-completion-function))
      (while (pcomplete-here (pcomplete-entries)))))

  (setq eshell-cmpl-cycle-completions nil)
  (setq tooltip-delay 3)
  ;;tab key fix
  (setq local-function-key-map (delq '(kp-tab . [9]) local-function-key-map))

  (use-package smartparens
  :ensure t
  :config
  (smartparens-global-mode 1)
  )

  (use-package openwith
  :ensure t
  :config
  (openwith-mode t)
  (setq openwith-associations '(("\\.pdf\\'" "evince" (file))
                                ("\\.docx\\'" "libreoffice" (file))
                                ("\\.odp\\'" "libreoffice" (file))
                                ("\\.odt\\'" "libreoffice" (file))
                                ("\\.MOV\\'" "totem" (file))
  ))
  )

;; prevent <openwith> from interfering with mail attachments
(require 'mm-util)
(add-to-list 'mm-inhibit-file-name-handlers 'openwith-file-handler)
#+END_SRC

* speed up startup
#+BEGIN_SRC emacs-lisp
;;; Temporarily reduce garbage collection during startup. Inspect `gcs-done'.
(defun ambrevar/reset-gc-cons-threshold ()
  (setq gc-cons-threshold (car (get 'gc-cons-threshold 'standard-value))))
(setq gc-cons-threshold (* 64 1024 1024))
(add-hook 'after-init-hook #'ambrevar/reset-gc-cons-threshold)

;;; Temporarily disable the file name handler.
(setq default-file-name-handler-alist file-name-handler-alist)
(setq file-name-handler-alist nil)
(defun ambrevar/reset-file-name-handler-alist ()
  (setq file-name-handler-alist default-file-name-handler-alist))
(add-hook 'after-init-hook #'ambrevar/reset-file-name-handler-alist)

#+END_SRC

* key chord
#+BEGIN_SRC emacs-lisp
(use-package key-chord
:ensure t
:config
(key-chord-mode 1)
)
#+END_SRC
* Python-elpy
** My pdb functions
- in order to make these functions work I need to import all of python mode.. Its ridiculous
#+BEGIN_SRC emacs-lisp
  (use-package python-mode
       :ensure t



  )

  (require 'pdb_emacs.el)

#+END_SRC
** initialize elpy
#+BEGIN_SRC emacs-lisp
  (use-package elpy
  :ensure t
  :config
  (elpy-enable)
  (setenv "IPY_TEST_SIMPLE_PROMPT" "1")
  (setq python-shell-interpreter "ipython3"
      ;; python-shell-interpreter-args "--profile=my_pylab"
                                          )
  (key-chord-define-global ".3" 'ipython3)

  (defun my-elpy-hook ()
      (define-key my-keys-minor-mode-map (kbd "C-S-i") nil)
      (define-key my-keys-minor-mode-map (kbd "C-S-k") nil)
      (local-set-key (kbd "M-t") 'elpy-nav-move-line-or-region-up)
      (local-set-key (kbd "M-w") 'elpy-nav-move-line-or-region-down)
      (local-set-key (kbd "M-r") 'elpy-nav-indent-shift-right)
      (local-set-key (kbd "M-h") 'elpy-nav-indent-shift-left)
      (local-set-key (kbd "C-c p") 'my-insert-pdb-break)
      (local-set-key (kbd "C-c o") 'my-pdb-cleanup)
      (local-set-key (kbd "C-c C-/")    'my-increment-number-decimal)
      (local-set-key (kbd "C-c C-l")    'my-decrement-number-decimal)
      (setq elpy-disable-backend-error-display 1)
      )
  (add-hook 'elpy-mode-hook 'my-elpy-hook)


#+END_SRC
* ipython settings
#+BEGIN_SRC emacs-lisp

   (defun my-inf-py-hook ()
          (setq kill-buffer-query-functions (delq 'process-kill-buffer-query-function kill-buffer-query-functions))
          (local-set-key "\C-cp" 'pdb-pm)
          )
   (add-hook 'inferior-python-mode-hook 'my-inf-py-hook))

  (defun eshell/sp ()
    (setq python-shell-interpreter "isympy3")
    (setq python-shell-interpreter-args "--pretty=unicode -I --")
    (run-python)
    (python-shell-switch-to-shell)
    (delete-other-windows)
    )
  (use-package ob-ipython
    :ensure t)

#+END_SRC
* Org-mode
** config
#+BEGIN_SRC emacs-lisp
    (use-package org
    :bind("C-c a". org-agenda)
    :config
    (setq org-log-done t)
    (global-set-key (kbd "C-c l") 'org-store-link)
    (setq org-capture-templates
     '(("T" "todo with link" entry (file "~/org/agenda/todo.org")
        "* TODO %i%? \n:PROPERTIES: \n:CREATED: %U \n:END: \n %a\n")
       ("t" "todo" entry (file "~/org/agenda/todo.org")
        "* TODO %i%? \n:PROPERTIES: \n:CREATED: %U \n:END: \n")
       ("b" "stuff to buy" entry (file+headline "~/org/agenda/buy.org" "to buy")
        "* TODO buy %i%?  %^g \n:PROPERTIES: \n:CREATED: %U \n:END: \n")
       ("g" "groceries to buy" entry (file+headline "~/org/agenda/buy.org" "groceries")
        "* TODO %i%? \n:PROPERTIES: \n:CREATED: %U \n:END: \n")
       ("B" "elfeed bookmark" entry (file+headline "~/org/bookmarks.org" "elfeed")
        "* %i%? \n:PROPERTIES: \n:CREATED: %U \n:END: \n %a")
       ("s" "school" entry
        (file "~/org/agenda/school.org")
        "")
       ("w" "work" entry
        (file "~/org/agenda/work.org")
        "")
       ("h" "home" entry
        (file "~/org/agenda/home.org")
        "")
       ("l" "lab" entry
        (file "~/org/agenda/lab.org")
        "")
       ("a" "Appointment" entry (file "~/org/agenda/appointments.org")
          "* %? %^G \n  %^t")
       ("e" "todo email" entry (file+headline "~/org/agenda/todo.org" "emails")
        "* TODO %i%? \n:PROPERTIES: \n:CREATED: %U \n:END: \n %a\n")
       ))

    (setq org-capture-templates-contexts
        '(("B" ((in-mode . "elfeed-show-mode")))
          ("e" ((in-mode . "mu4e-view-mode")))))

    (setq org-agenda-files (list "~/org/agenda"))
    (setq org-archive-location "~/org/agenda/archive.org::* From %s")
    (require 'org-mu4e)
    (add-hook 'org-mode-hook (lambda () (modify-syntax-entry (string-to-char "\u25bc") "w"))) ; Down arrow for collapsed drawer.
    (setq org-todo-keywords
        '((sequence "TODO" "WAITING" "|" "DONE" )))
    (setq org-startup-indented t)
    (setq org-hide-leading-stars t)
    (setq org-odd-level-only nil)
    (setq org-insert-heading-respect-content nil)
    (setq org-M-RET-may-split-line '((item) (default . t)))
    (setq org-special-ctrl-a/e t)
    (setq org-return-follows-link nil)
    (setq org-use-speed-commands t)
    (setq org-startup-align-all-tables nil)
    (setq org-log-into-drawer nil)
    (setq org-tags-column 1)
    (setq org-ellipsis " \u25bc" )
    (setq org-speed-commands-user nil)
    (setq org-blank-before-new-entry '((heading . nil) (plain-list-item . nil)))
    (setq org-completion-use-ido t)
    (setq org-indent-mode t)
    (setq org-startup-truncated nil)
    (setq auto-fill-mode -1)
    (setq-default fill-column 99999)
    (setq fill-column 99999)
    (setq org-image-actual-width nil)
    (setq org-confirm-babel-evaluate nil)
    (add-hook 'org-babel-after-execute-hook 'org-display-inline-images 'append)
    (global-auto-revert-mode t)
    (prefer-coding-system 'utf-8)
    ;; (setq cua-auto-tabify-rectangles nil) ;; Don't tabify after rectangle commands
    (transient-mark-mode nil)               ;; No region when it is not highlighted
    ;; (setq cua-keep-region-after-copy t)
    (global-set-key (kbd "C-c c") 'org-capture)
    (org-babel-do-load-languages
     'org-babel-load-languages
    '((python . t)
      (ipython . t)
      ))

  )



  (defun org-override-keys ()
    (local-set-key (kbd "C-/") 'comment-or-uncomment-line-or-region)
    (local-set-key (kbd "M-h") 'org-shiftmetaleft)
    (local-set-key (kbd "M-r") 'org-shiftmetaright)
    )
    (add-hook 'org-mode-hook 'org-override-keys)

    (defun my-org-agenda-hook ()
            (local-set-key (kbd "C-t") 'org-agenda-previous-line)
            (local-set-key (kbd "C-w") 'org-agenda-next-line)
                   )
    (add-hook 'org-agenda-mode-hook 'my-org-agenda-hook)

  (use-package org-bullets
        :ensure t
        :config
        (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1))))


#+END_SRC
** latex classes

* themes
#+BEGIN_SRC emacs-lisp

(add-to-list 'custom-theme-load-path "~/.emacs.d/themes")
(load-theme 'tron-legacy t)

#+END_SRC
* multiple cursors

config.el

;; functions
(defun backward-kill-line ()
  "Kill a backward line"
  (interactive)
  (kill-line -1)
  )

(defun comment-or-uncomment-line-or-region ()
  "Comments or uncomments the current line or region."
  (interactive)
  (if (region-active-p)
      (comment-or-uncomment-region (region-beginning) (region-end))
    (comment-or-uncomment-region (line-beginning-position) (line-end-position))
    )
  )

(defun kill-line-or-region ()
  "kills current line or region."
  (interactive)
  (if (region-active-p)
      (kill-region (region-beginning)(region-end))
    (kill-region (line-beginning-position) (line-end-position))
    )
  )

(defun my-insert-date (prefix)
    "Insert the current date. With prefix-argument, use ISO format. With
   two prefix arguments, write out the day and month name."
    (interactive "P")
    (let ((format (cond
                   ((not prefix) "%Y-%m-%d %H:%M") 
                   ((equal prefix '(4)) "%Y-%m-%d")
                   ((equal prefix '(16)) "%A, %d. %B %Y")))
          )
      (insert (format-time-string format))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;               insert a file name
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defun my-insert-file-name (filename &optional args)
    "Insert name of file FILENAME into buffer after point.
  
  Prefixed with \\[universal-argument], expand the file name to
  its fully canocalized path.  See `expand-file-name'.
  
  Prefixed with \\[negative-argument], use relative path to file
  name from current directory, `default-directory'.  See
  `file-relative-name'.
  
  The default with no prefix is to insert the file name exactly as
  it appears in the minibuffer prompt."
    ;; Based on insert-file in Emacs -- ashawley 20080926
    (interactive "*fInsert file name: \nP")
    (cond ((eq '- args)
           (insert (file-relative-name filename)))
          ((not (null args))
           (insert (expand-file-name filename)))
          (t
           (insert filename))))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;Cleanup commands;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defun untabify-buffer ()
  "Untabify current buffer"
  (interactive)
  (untabify (point-min) (point-max)))

(defun indent-buffer()
  "indent whole buffer"
  (interactive)
  (delete-trailing-whitespace)
  (indent-region (point-min) (point-max) nil)
  (untabify (point-min) (point-max)))

(defun cleanup-buffer ()
  "Perform a bunch of operations on the whitespace content of a buffer."
  (interactive)
  (indent-buffer)
  (untabify-buffer)
  (delete-trailing-whitespace))

(global-set-key (kbd "C-i") 'previous-line)
(global-set-key (kbd "C-k") 'next-line)
(global-set-key (kbd "C-l") 'forward-char)
(global-set-key (kbd "C-j") 'backward-char)
(global-set-key (kbd "C-S-i") 'scroll-down)
(global-set-key (kbd "C-S-k") 'scroll-up)
(global-set-key (kbd "C-S-l") 'forward-word)
(global-set-key (kbd "C-S-j") 'backward-word)
(global-set-key (kbd "C-x C-l") 'end-of-line)
(global-set-key (kbd "C-x C-j") 'beginning-of-line)
(global-set-key (kbd "C-x C-k") 'end-of-buffer)
(global-set-key (kbd "C-x C-i") 'beginning-of-buffer)

(global-set-key (kbd "M-d") 'other-window)
(global-set-key (kbd "C-o") 'split-window-right)
(global-set-key (kbd "C-S-o") 'split-window-below)
(global-set-key (kbd "C-x C-o") 'delete-window)

(global-set-key (kbd "C-b") 'switch-to-next-buffer)
(global-set-key (kbd "C-x C-b") 'switch-to-buffer)
(global-set-key (kbd "M-b") 'switch-to-prev-buffer)

(global-set-key (kbd "M-i") 'backward-kill-line)
(global-set-key (kbd "M-k") 'kill-line)
(global-set-key (kbd "M-l") 'delete-forward-char)
(global-set-key (kbd "M-j") 'delete-backward-char)
(global-set-key (kbd "M-S-i") 'previous-line)
(global-set-key (kbd "M-S-k") 'next-line)
(global-set-key (kbd "M-L") 'kill-word)
(global-set-key (kbd "M-J") 'backward-kill-word)

(global-set-key (kbd "C-s--") 'copy-region-as-kill)
(global-set-key (kbd "C-=") 'yank)
(local-set-key (kbd "C-s-=") 'yank-pop)
(global-set-key (kbd "C--") 'kill-line-or-region)
(global-set-key (kbd "C-z") 'undo)

(global-set-key (kbd "M-q") 'keyboard-escape-quit)

(global-set-key (kbd "C-/") 'comment-or-uncomment-line-or-region)

(global-set-key (kbd "C-x C-,") 'kmacro-start-macro)
(global-set-key (kbd "C-x C-.") 'kmacro-end-macro)
(global-set-key (kbd "C-x C-/") 'kmacro-end-and-call-macro)

(global-set-key "\C-c i d"    'my-insert-date)
(global-set-key "\C-c i f" 'my-insert-file-name)
(global-set-key "\C-c n"    'cleanup-buffer)

(defvar my-keys-minor-mode-map (make-keymap) "my-keys-minor-mode keymap.")
(define-key my-keys-minor-mode-map (kbd "C-i") 'previous-line)
(define-key my-keys-minor-mode-map (kbd "C-j") 'backward-char)
(define-key my-keys-minor-mode-map (kbd "C-i") 'previous-line)
(define-key my-keys-minor-mode-map (kbd "C-k") 'next-line)
(define-key my-keys-minor-mode-map (kbd "C-l") 'forward-char)
(define-key my-keys-minor-mode-map (kbd "C-j") 'backward-char)
(define-key my-keys-minor-mode-map (kbd "C-S-i") 'scroll-down)
(define-key my-keys-minor-mode-map (kbd "C-S-k") 'scroll-up)
(define-key my-keys-minor-mode-map (kbd "C-S-i") 'backward-paragraph)
(define-key my-keys-minor-mode-map (kbd "C-S-k") 'forward-paragraph)
(define-key my-keys-minor-mode-map (kbd "C-S-l") 'forward-word)
(define-key my-keys-minor-mode-map (kbd "C-S-j") 'backward-word)

(define-key my-keys-minor-mode-map (kbd "M-d") 'other-window)
(define-key my-keys-minor-mode-map (kbd "C-o") 'split-window-right)
(define-key my-keys-minor-mode-map (kbd "C-S-o") 'split-window-below)
(define-key my-keys-minor-mode-map (kbd "C-x C-o") 'delete-window)

(define-key my-keys-minor-mode-map (kbd "C-b") 'switch-to-next-buffer)
(define-key my-keys-minor-mode-map (kbd "C-x C-b") 'switch-to-buffer)

;;killing
(define-key my-keys-minor-mode-map (kbd "M-i") 'backward-kill-line)
(define-key my-keys-minor-mode-map (kbd "M-k") 'kill-line)
(define-key my-keys-minor-mode-map (kbd "M-l") 'delete-forward-char)
(define-key my-keys-minor-mode-map (kbd "M-j") 'delete-backward-char)
(define-key my-keys-minor-mode-map (kbd "M-S-i") 'previous-line)
(define-key my-keys-minor-mode-map (kbd "M-S-k") 'next-line)
(define-key my-keys-minor-mode-map (kbd "M-S-l") 'kill-word)
(define-key my-keys-minor-mode-map (kbd "M-S-j") 'backward-kill-word)

;; copy,cut,paste,undo
(define-key my-keys-minor-mode-map (kbd "C-s--") 'copy-region-as-kill)
(define-key my-keys-minor-mode-map (kbd "C-=") 'yank)
(define-key my-keys-minor-mode-map (kbd "C-s-=") 'yank-pop)
(define-key my-keys-minor-mode-map (kbd "C--") 'kill-line-or-region)
(define-key my-keys-minor-mode-map (kbd "C-z") 'undo)

;;keyboard quit
(define-key my-keys-minor-mode-map (kbd "M-q") 'keyboard-escape-quit)

;;commenting
(define-key my-keys-minor-mode-map (kbd "C-/") 'comment-or-uncomment-line-or-region)

(define-minor-mode my-keys-minor-mode
  "A minor mode so that my key settings override annoying major modes."
  t " my-keys" 'my-keys-minor-mode-map)

;;;;;;;;;on startup;;;;;;;;
(my-keys-minor-mode 1)

;;;;;;;;everything else ;;;;;;;

(defun my-minibuffer-setup-hook ()
  (my-keys-minor-mode 0))
(add-hook 'minibuffer-setup-hook 'my-minibuffer-setup-hook)


(defun my-erc-setup-hook ()
  (my-keys-minor-mode 0))
(add-hook 'erc-mode-hook 'my-erc-setup-hook)

(menu-bar-mode -1) 
(toggle-scroll-bar -1) 
(tool-bar-mode -1)

(electric-indent-mode 1)

;;dont have to use "yes" or "no" instead use "y" or "n"
(defalias 'yes-or-no-p 'y-or-n-p)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;;;;;;;;;;;;;;; emacs autocomplete with tab;;;;;;;;;;;
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(add-hook
 'eshell-mode-hook
 (lambda ()
   (setq pcomplete-cycle-completions nil)))

(setq eshell-cmpl-cycle-completions nil)

;;tab key fix
(setq local-function-key-map (delq '(kp-tab . [9]) local-function-key-map))

(smartparens-global-mode 1)

(add-hook 'after-init-hook 'global-flycheck-mode)

(defun my-python-send-region (&optional beg end)
  (interactive)
  (let ((beg (cond (beg beg)
                   ((region-active-p)
                    (region-beginning))
                   (t (line-beginning-position))))
        (end (cond (end end)
                   ((region-active-p)
                    (copy-marker (region-end)))
                   (t (line-end-position)))))
    (py-execute-region-ipython-no-switch beg end)))

(use-package python-mode 
:config 
(setq
 python-shell-interpreter "ipython"
 python-shell-interpreter-args ""
 python-shell-prompt-regexp "In \\[[0-9]+\\]: "
 python-shell-prompt-output-regexp "Out\\[[0-9]+\\]: "
 python-shell-completion-setup-code
   "from IPython.core.completerlib import module_completion"
 python-shell-completion-module-string-code
   "';'.join(module_completion('''%s'''))\n"
 python-shell-completion-string-code
   "';'.join(get_ipython().Completer.all_completions('''%s'''))\n"))

(defun python-override-keys ()
(local-set-key (kbd "C-c C-c") 'my-python-send-region)
(local-set-key (kbd "C-c C-s") 'py-execute-buffer-ipython-no-switch)
(local-set-key (kbd "M-RET") 'py-newline-and-indent)

(local-set-key (kbd "M-o") 'py-shift-right)
(local-set-key (kbd "M-u") 'py-shift-left)
)

(add-hook 'python-mode-hook 'python-override-keys)
(add-hook 'py-ipython-shell-mode-hook 'python-override-keys)

(add-hook 'python-mode-hook 'jedi:setup)
(setq jedi:complete-on-dot t) 
(setq jedi:get-in-function-call-delay 500)
(setq jedi:get-in-function-call-timeout 1000)

(use-package org
:bind("C-c a". org-agenda)
:config
(setq org-log-done t)
(setq org-agenda-files (list "~/org/work.org"
                             "~/org/school.org" 
                             "~/org/home.org"))
(add-hook 'org-mode-hook (lambda () (modify-syntax-entry (string-to-char "\u25bc") "w"))) ; Down arrow for collapsed drawer.
(setq org-startup-indented t)
(setq org-hide-leading-stars t)
(setq org-odd-level-only nil) 
(setq org-insert-heading-respect-content nil)
(setq org-M-RET-may-split-line '((item) (default . t)))
(setq org-special-ctrl-a/e t)
(setq org-return-follows-link nil)
(setq org-use-speed-commands t)
(setq org-startup-align-all-tables nil)
(setq org-log-into-drawer nil)
(setq org-tags-column 1)
(setq org-ellipsis " \u25bc" )
(setq org-speed-commands-user nil)
(setq org-blank-before-new-entry '((heading . nil) (plain-list-item . nil)))
(setq org-completion-use-ido t)
(setq org-indent-mode t)
(setq org-startup-truncated nil)
(setq auto-fill-mode -1)
(setq-default fill-column 99999)
(setq fill-column 99999)
(global-auto-revert-mode t)
(prefer-coding-system 'utf-8)
;; (setq cua-auto-tabify-rectangles nil) ;; Don't tabify after rectangle commands
(transient-mark-mode nil)               ;; No region when it is not highlighted
;; (setq cua-keep-region-after-copy t) 
)



(defun org-override-keys ()
(local-set-key (kbd "C-/") 'comment-or-uncomment-line-or-region)
(local-set-key (kbd "M-u") 'org-metaleft)
(local-set-key (kbd "M-o") 'org-metaright)
)

(add-hook 'org-mode-hook 'org-override-keys)

(add-to-list 'custom-theme-load-path "~/.emacs.d/themes")
(load-theme 'zenburn t)

(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(load custom-file)

(use-package key-chord
:ensure t
:config
(key-chord-mode 1)
(key-chord-define-global "o0" 'find-file)
(key-chord-define-global "w2" 'ipython)
)

(use-package arduino-mode)

(load "auctex.el" nil t t)
(setq TeX-auto-save t)
(setq TeX-parse-self t)
(setq-default TeX-master nil)

(add-hook 'LaTeX-mode-hook 'turn-on-reftex)
(setq reftex-plug-into-AUCTeX t)
;; So that RefTeX finds my bibliography
(setq reftex-default-bibliography '("/home/nick/my_zotero.bib"))

(add-hook 'TeX-mode-hook 'zotelo-minor-mode)

;;; auto complete mod
;;; should be loaded after yasnippet so that they can work together
(use-package auto-complete-config
:config
(add-to-list 'ac-dictionary-directories "~/.emacs.d/ac-dict")
(ac-config-default))


(use-package yasnippet
:ensure t
:config
(yas-global-mode 1)

(define-key yas-minor-mode-map [(tab)] nil)
(define-key yas-minor-mode-map (kbd "TAB") nil)
(define-key yas-minor-mode-map (kbd "C-x C-y") 'yas-expand)
;;; set the trigger key so that it can work together with yasnippet on tab key,
;;; if the word exists in yasnippet, pressing tab will cause yasnippet to
;;; activate, otherwise, auto-complete will
(ac-set-trigger-key "TAB")
(ac-set-trigger-key "<tab>"))
